Improved generic exception "Error reading non-2xx response body"

tidoemanuele opened the pull request:
This will allow the **onPlayerError** method to get the response http status code, message and headers also when there is an error getting the **errorResponseBody**

tidoemanuele committed:
throw InvalidResponseCodeException instead of HttpDataSourceException

if responseCode and responseMessage ara available always throws an InvalidResponseCodeException instead of HttpDataSourceException, so in onPlayerError method the http status code and message can be used to decide what will be the next step.

tidoemanuele committed:
throw exception improved

if responseCode and responseMessage ara available always throws an InvalidResponseCodeException instead of HttpDataSourceException, so in onPlayerError method the http status code and message can be used to decide what will be the next step.

tidoemanuele committed:
Update DefaultHttpDataSource.java

ojw28 commented on extensions/okhttp/src/main/java/com/google/android/exoplayer2/ext/okhttp/OkHttpDataSource.java:
```diff
@@ -304,14 +304,14 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
       } catch (IOException e) {
-        throw new HttpDataSourceException(
-            "Error reading non-2xx response body", e, dataSpec, HttpDataSourceException.TYPE_OPEN);
+        throw new InvalidResponseCodeException(
```
Isn't it better just to replace this with a `errorResponseBody = Util.EMPTY_BYTE_ARRAY` and then fall through, so that the lines below take care of throwing what you want? I think that should just work. Note that you need to do it this way for the special handling of `416` to work as well.

Note that it will not be possible to disambiguate between "no response body" and "failed to read response body" in this case. If you think it's important to be able to disambiguate then you can use `errorResponseBody = null` instead. I'm not sure it's worth it though?

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/upstream/DefaultHttpDataSource.java:
```diff
@@ -377,8 +377,8 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
       } catch (IOException e) {
-        throw new HttpDataSourceException(
-            "Error reading non-2xx response body", e, dataSpec, HttpDataSourceException.TYPE_OPEN);
+        throw new InvalidResponseCodeException(
```
Same comment as in `OkHttpDataSource`. I think it's better just to set `errorResponseBody` to Util.EMPTY_BYTE_ARRAY` and fall through.

Note that `CronetDataSource` should also be modified as part of this change. Probably by enclosing [this while loop](https://github.com/google/ExoPlayer/blob/1347d572ef9ce79aacd667cfffa7d7468c8408a4/extensions/cronet/src/main/java/com/google/android/exoplayer2/ext/cronet/CronetDataSource.java#L555) in a `try/catch` block and then in the `catch` block setting `responseBody` back to `Util.EMPTY_BYTE_ARRAY` before falling through.

ojw28 commented on extensions/okhttp/src/main/java/com/google/android/exoplayer2/ext/okhttp/OkHttpDataSource.java:
```diff
@@ -304,14 +304,14 @@ public long open(DataSpec dataSpec) throws HttpDataSourceException {
-        throw new HttpDataSourceException(
-            "Error reading non-2xx response body", e, dataSpec, HttpDataSourceException.TYPE_OPEN);
+        throw new InvalidResponseCodeException(
+            responseCode, response.message(), headers, dataSpec, null);
```
The `responseBody` argument isn't `@Nullable`, so you should use `Util.EMPTY_BYTE_ARRAY` as suggested above. Alternatively, you'll need to add `@Nullable` annotations through `InvalidResponseCodeException`. I think it's probably best to use `Util.EMPTY_BYTE_ARRAY` unless you can think of a case where it's important to be able to disambiguate between the cases described above.

ojw28 commented:
Please see comments inline. Thanks!

tidoemanuele committed:
Update DefaultHttpDataSource.java

set responseBody to Util.EMPTY_BYTE_ARRAY when an IOException occurs trying to get his value from inputStream

tidoemanuele committed:
Update OkHttpDataSource.java

set responseBody to Util.EMPTY_BYTE_ARRAY when an IOException occurs trying to get his value from inputStream

tidoemanuele committed:
Update DefaultHttpDataSource.java

tidoemanuele committed:
Update OkHttpDataSource.java

tidoemanuele reviewed (commented):
you are right, better to set the responseBody to empty there
